core/schema/modulesource.go

```diff
@@ -1491,9 +1523,306 @@ func (s *moduleSourceSchema) moduleSourceWithBlueprint(
+			for updateReq := range updateReqs {
+				if updateReq.symbolic == existingToolchain.Self().ModuleName {
+					return inst, fmt.Errorf("updating local toolchains is not supported")
+				}
+
+				var contextRoot string
+				switch parentSrc.Self().Kind {
+				case core.ModuleSourceKindLocal:
+					contextRoot = parentSrc.Self().Local.ContextDirectoryPath
+				case core.ModuleSourceKindGit:
+					contextRoot = "/"
+				default:
+					return inst, fmt.Errorf("unknown module source kind: %s", parentSrc.Self().Kind)
+				}
+
+				parentSrcRoot := filepath.Join(contextRoot, parentSrc.Self().SourceRootSubpath)
+				tcSrcRoot := filepath.Join(contextRoot, existingToolchain.Self().SourceRootSubpath)
+				existingSymbolic, err := pathutil.LexicalRelativePath(parentSrcRoot, tcSrcRoot)
+				if err != nil {
+					return inst, fmt.Errorf("failed to get relative path: %w", err)
+				}
+
+				if updateReq.symbolic == existingSymbolic {
+					return inst, fmt.Errorf("updating local toolchains is not supported")
+				}
+			}
+			continue
+		}
+
+		existingName := existingToolchain.Self().ModuleName
+		existingVersion := existingToolchain.Self().Git.Version
+		existingSymbolic := existingToolchain.Self().Git.CloneRef
+		if tcSrcRoot := existingToolchain.Self().SourceRootSubpath; tcSrcRoot != "" {
+			existingSymbolic += "/" + strings.TrimPrefix(tcSrcRoot, "/")
+		}
+
+		for updateReq := range updateReqs {
+			if updateReq.symbolic != existingName && updateReq.symbolic != existingSymbolic {
+				continue
+			}
+			delete(updateReqs, updateReq)
+
+			updateVersion := updateReq.version
+			if updateVersion == "" {
+				updateVersion = existingVersion
+			}
+			updateRef := existingSymbolic
+			if updateVersion != "" {
+				updateRef += "@" + updateVersion
+			}
+
+			var updatedToolchain dagql.ObjectResult[*core.ModuleSource]
+			err := dag.Select(ctx, dag.Root(), &updatedToolchain,
+				dagql.Selector{
+					Field: "moduleSource",
+					Args: []dagql.NamedInput{
+						{Name: "refString", Value: dagql.String(updateRef)},
+					},
+				},
+			)
+			if err != nil {
+				return inst, fmt.Errorf("failed to load updated toolchain: %w", err)
+			}
+
+			newUpdatedToolchainArgs = append(newUpdatedToolchainArgs, dagql.NewID[*core.ModuleSource](updatedToolchain.ID()))
+		}
+	}
+
+	if len(updateReqs) > 0 {
+		toolchains := make([]string, 0, len(updateReqs))
+		for updateReq := range updateReqs {
+			toolchains = append(toolchains, updateReq.symbolic)
+		}
+		return inst, fmt.Errorf("toolchain %q was requested to be updated, but it is not found in the toolchains list", strings.Join(toolchains, ","))
+	}
+
+	// Build a new module source with updated toolchains
+	tmpSrc := parentSrc.Self().Clone()
+	tmpSrc.Toolchains = nil
+	tmpSrc.ConfigToolchains = nil
+
+	for _, toolchainID := range newUpdatedToolchainArgs {
+		var toolchainSrc dagql.ObjectResult[*core.ModuleSource]
+		err := dag.Select(ctx, dag.Root(), &toolchainSrc,
+			dagql.Selector{
+				Field: "loadID",
```

Comment:
Don't think that's a thing 🤔 (though I wish it was)

```suggestion
				Field: "loadModuleSourceFromID",
```

Now that I look deeper though, is there a reason we need to re-load all these? Seems like we have a `dagql.ObjectResult[*core.ModuleSource]` in each case already before adding IDs to an array, only to re-load them - couldn't we just store the objects directly?

Reply:
Fixed, that must have been something missed in a refactor. I dont know what "loadID" was supposed to be lol